src/overlay.rs

```diff
@@ -0,0 +1,59 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+
+use crate::{
+    storage::{TreeReader, TreeWriter},
+    JellyfishMerkleTree, KeyHash, OwnedValue, RootHash, Version,
+};
+
+// There's some similarity between the OverlayTree and the TreeCache, but it's
+// not exactly clear how to share code between them (the TreeCache operates at
+// the level of the tree internals, not at the level of its external interface),
+// so it's easier and better for now to just make a new wrapper.
+
+pub struct OverlayTree<'a, R> {
+    tree: JellyfishMerkleTree<'a, R>,
+    overlay: HashMap<KeyHash, OwnedValue>,
+    version: Version,
+}
```

Comment:
A better design here would be to do
```suggestion
pub struct OverlayTree<R> {
    reader: R,
    overlay: HashMap<KeyHash, OwnedValue>,
    version: Version,
}
```
and instead of using `self.tree`, just call methods on `JellyfishMerkleTree(&self.reader)`.

The benefit is that by having the `OverlayTree` own the reader, there's no need to manage the reader lifetime, and the overlay tree can be `'static`.